feat: add auto-checkout for missing employee check-outs

Nikhil-Tummala opened the pull request:
## 🚀 Feature: Auto-checkout for Missing Employee Check-outs

### 📋 Description
This PR implements an automated solution for issue #3353, ensuring HRMS always shows the correct "Check In" option for employees starting a new day.

**Closes #3353**

### 💡 Problem (from #3353)
- When employees forget to check out on a certain day, HRMS shows "Check Out" option the next day
- This creates confusion as employees expect to see "Check In" when starting a new workday
- Manual intervention required to fix incomplete attendance records

### ✨ Solution Implemented
- **Auto-checkout Function**: Automatically creates check-out entries at 23:59:59 for employees who checked in but didn't check out the previous day
- **Daily Scheduled Task**: Runs automatically to process missing check-outs
- **Smart Detection**: Only processes genuine missing check-outs, prevents duplicates
- **Result**: Employees always see "Check In" option when starting a new day

### 🔧 Technical Changes
- Added `auto_checkout_missing_checkouts()` function in `hrms/hr/doctype/employee_checkin/employee_checkin.py`
- Added daily scheduler event in `hrms/hooks.py`

### 🧪 Testing Scenarios
- [x] Employee checks in but forgets to check out → Auto check-out created at 23:59:59
- [x] Employee with complete check-in/check-out pair → No changes made
- [x] Prevents duplicate check-out creation
- [x] Next day correctly shows "Check In" option

**Fixes #3353**

Nikhil-Tummala committed:
feat: auto-checkout for missing employee check-outs

- Add auto_checkout_missing_checkouts function to automatically create check-out entries for employees who checked in but forgot to check out the previous day
- Automatically creates check-out at 23:59:59 for missing check-outs from yesterday
- Prevents duplicate check-out creation by checking for existing check-out entries
- Add scheduled task to run daily check for missing check-outs
- Improves attendance tracking accuracy by handling incomplete check-in/check-out pairs
- Ensures HRMS always shows Check In option for new day instead of Check Out

Fixes #3353

asmitahase commented on hrms/hr/doctype/employee_checkin/employee_checkin.py:
```diff
@@ -397,3 +397,37 @@ def update_attendance_in_checkins(log_names: list, attendance_id: str):
+                "time": ["between", [str(yesterday) + " 00:00:00", str(yesterday) + " 23:59:59"]],
+            }
+        )
+        if not out_exists:
+            # Create a check-out at 23:59:59
+            doc = frappe.new_doc("Employee Checkin")
+            doc.employee = checkin.employee
+            doc.log_type = "OUT"
+            doc.time = str(yesterday) + " 23:59:59"
```
I think a better way would be to add the checkout near the shift end time for the employee's shift?
Creating checkout at midnight may cause it to be outside of the shift time and end up not being considered for auto attendance.

Nikhil-Tummala replied:
@asmitahase Thank you for the feedback! You're absolutely right, and I've already implemented this enhancement. The current implementation now:

1. **Uses shift end time instead of midnight** - The function now calculates checkout time based on the employee's actual shift assignment using `get_actual_start_end_datetime_of_shift()`

2. **Respects shift timings** - If an employee has a shift assigned and auto-checkout is enabled for that shift type, it uses the `end_datetime` from the shift (not `actual_end` which includes grace periods)

3. **Falls back gracefully** - If no shift is assigned or shift auto-checkout is disabled, it uses the configurable max shift duration from HR Settings

4. **Added configuration options**:
   - **HR Settings**: `enable_auto_checkout` (global toggle) and `max_shift_duration` (fallback duration)
   - **Shift Type**: `enable_auto_checkout` (per-shift control)

The logic now works as:
```python
# If shift exists and has auto-checkout enabled
if shift_timings and shift_timings.shift_type.enable_auto_checkout:
    shift_end = shift_timings.end_datetime
    max_time = checkin_time + timedelta(hours=max_shift_duration)
    checkout_time = min(shift_end, max_time)  # Use earlier of shift end or max duration

asmitahase commented on hrms/hooks.py:
```diff
@@ -232,6 +232,7 @@
 		"hrms.hr.doctype.daily_work_summary_group.daily_work_summary_group.send_summary",
 		"hrms.hr.doctype.interview.interview.send_daily_feedback_reminder",
 		"hrms.hr.doctype.job_opening.job_opening.close_expired_job_openings",
+		"hrms.hr.doctype.employee_checkin.employee_checkin.auto_checkout_missing_checkouts",
```
This should be configurable ideally through shift type or at least through HR Settings

mahsem replied:
@asmitahase @Nikhil-Tummala maybe add max shift duration time per employee and check against that value

Nikhil-Tummala committed:
feat: Add configurable auto-checkout functionality for missing employee checkouts

- Add auto_checkout_missing_checkouts() function in employee_checkin.py
  * Uses shift end time instead of midnight for accurate checkout timing
  * Respects shift assignments and shift-specific auto-checkout settings
  * Falls back to max shift duration when shift auto-checkout is disabled
  * Logs errors and adds comments to explain auto-generated checkouts

- Add HR Settings configuration fields:
  * enable_auto_checkout: Global toggle for auto-checkout feature
  * max_shift_duration: Maximum hours for shift duration (default: 8)

- Add Shift Type configuration field:
  * enable_auto_checkout: Per-shift toggle for granular control

- All features are disabled by default to ensure backward compatibility
- Auto-checkouts are marked with skip_auto_attendance=1 to avoid interference
- Proper error handling and logging for production reliability

Addresses PR feedback: uses shift timing logic instead of fixed midnight checkout

asmitahase commented:
This looks like vibe-coded PR to me. The auto checkout function is way too long. New fields are added directly to the json instead of modifying through desk. Closing this for now. Will revisit this later
